ttnn/cpp/ttnn/graph_processor.cpp

```diff
@@ -427,16 +427,47 @@ namespace ttnn {
             .params = {},
             .connections = {}
         });
-        hooks = std::make_shared<ProcessorHooks>();
-        tt::tt_metal::GraphTracker::instance().add_hook(hooks);
-        hooks->set_block(mode == RunMode::FAKE);
+
+        if (!tt::tt_metal::GraphTracker::instance().get_hook()) {
+            hooks = std::make_shared<ProcessorHooks>();
+            tt::tt_metal::GraphTracker::instance().add_hook(hooks);
+            hooks->set_block(mode == RunMode::FAKE);
+        }
         current_op_id.push(0);
     }
     nlohmann::json GraphProcessor::end_capture() {
         const std::lock_guard<std::mutex> lock(mutex);
+        int counter = graph.size();
+        graph.push_back(Vertex{
+            .counter = 0,
+            .name = "capture_end",
+            .params = {},
+            .connections = {}
+        });
+        if ( last_finished_op_id != -1 ) {
+            graph[last_finished_op_id].connections.push_back(counter);
+        } else {
+            // lets connect capture_start with capture_end
+            // it means we didn't capture any functions
+            TT_ASSERT(current_op_id.size(), "Graph size could not be 0. It means track_function_end called more then begin.");
```

Comment:
Nit: Grammar, and should `track_function_end` be `track_end_function`? `"Graph size cannot be 0. This means that track_end_function was called more than begin."`